Mathlib/RingTheory/FractionalIdeal/Extended.lean

```diff
@@ -0,0 +1,132 @@
+/-
+Copyright (c) 2024 James Sundstrom. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: James Sundstrom
+-/
+import Mathlib.RingTheory.FractionalIdeal.Basic
+
+/-!
+# Extension of fractional ideals
+
+This file defines the extension of a fractional ideal along a ring homomorphism.
+
+## Main definition
+
+* `FractionalIdeal.extended`: Let `A` and `B` be commutative rings with respective localizations
+  `IsLocalization M K` and `IsLocalization N L`. Let `f : A →+* B` be a ring homomorphism with
+  `hf : M ≤ Submonoid.comap f N`. If `I : FractionalIdeal M K`, then the extension of `I` along
+  `f` is `extended L hf I : FractionalIdeal N L`.
+
+## Main results
+
+* `extended_add` says that extension commutes with addition.
+* `extended_mul` says that extension commutes with multiplication.
+
+## Tags
+
+fractional ideal, fractional ideals, extended, extension
+-/
+
+open IsLocalization FractionalIdeal Submodule
+
+namespace FractionalIdeal
+
+variable {A : Type*} [CommRing A] {B : Type*} [CommRing B] {f : A →+* B}
+variable {K : Type*} {M : Submonoid A} [CommRing K] [Algebra A K] [IsLocalization M K]
+variable (L : Type*) {N : Submonoid B} [CommRing L] [Algebra B L] [IsLocalization N L]
+variable (hf : M ≤ Submonoid.comap f N)
+variable (I : FractionalIdeal M K) (J : FractionalIdeal M K)
+
+/-- Given commutative rings `A` and `B` with respective localizations `IsLocalization M K` and
+`IsLocalization N L`, and a ring homomorphism `f : A →+* B` satisfying `M ≤ Submonoid.comap f N`, a
+fractional ideal `I` of `A` can be extended along `f` to a fractional ideal of `B`. -/
+def extended (I : FractionalIdeal M K) : FractionalIdeal N L where
+  val := span B <| (IsLocalization.map (S := K) L f hf) '' I
+  property := by
+    have ⟨a, ha, frac⟩ := I.isFractional
+    refine ⟨f a, hf ha, fun b hb ↦ ?_⟩
+    refine span_induction hb (fun x hx ↦ ?_) ⟨0, by simp⟩
+      (fun x y hx hy ↦ smul_add (f a) x y ▸ isInteger_add hx hy) (fun b c hc ↦ ?_)
+    · rcases hx with ⟨k, kI, rfl⟩
+      obtain ⟨c, hc⟩ := frac k kI
+      exact ⟨f c, by simp [← IsLocalization.map_smul, ← hc]⟩
+    · rw [← smul_assoc, smul_eq_mul, mul_comm (f a), ← smul_eq_mul, smul_assoc]
+      exact isInteger_smul hc
+
+local notation "map_f" => (IsLocalization.map (S := K) L f hf)
+
+lemma mem_extended_iff (x : L) : (x ∈ I.extended L hf) ↔ x ∈ span B (map_f '' I) := by
+  constructor <;> { intro hx; simpa }
+
+lemma extended_eq : (I.extended L hf).coeToSubmodule = span B (map_f '' I) := by
+  ext; simp [mem_coe, mem_extended_iff]
+
+theorem extended_zero : extended L hf (0 : FractionalIdeal M K) = 0 := by
+  rw [extended, eq_zero_iff]
```

Comment:
```suggestion
theorem extended_zero : extended L hf (0 : FractionalIdeal M K) = 0 := by
  apply coeToSubmodule_injective
  simp only [coe_extended_eq_span, coe_zero]
  have : ((0 : FractionalIdeal M K) : Set K) = {0} := by ext; simp
  simp [this]
```
I think all of these can be streamlined a bit by reducing immediately to a question about submodules. 

`convert` and `change`s should probably a tool of last resort. I think they can all be removed. 

This could also be make a `simp`.